appsrc/dcfile/dcstats.man
.TH DCSTATS 1 "6 March 2007" "DICOM PS3" "DICOM PS3 - DICOM image statistics"
.SH NAME
dcstats \- ACR/NEMA DICOM PS3 ... DICOM PS3 - DICOM image statistics
.SH SYNOPSIS
.HP 10
.B dcstats
[
.B \-v|verbose
]
[
.B \-filename
]
.so man1/optin.so
.SH DESCRIPTION
.LP
.B dcstats
reads the named dicom or acr-nema input file and describes the statistics
of the image pixel data. The primary intent is to determine if the pixel
representation is unsigned or signed (regardless of what other attributes
may specify).
.LP
The description and verbose output go to standard error.
.PP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-v|verbose
.RS
Display byte offset from file start and message start, in hexadecimal, and dump contents while reading and once read.
.RE
.TP
.B \-filename
.RS
Show the name of the file supplied in the arguments; a hyphen will be reported if no filename was supplied.
.RE
.TP
.B \-filename
.RS
Show the name of the file supplied in the arguments; a hyphen will be reported if no filename was supplied.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% dcstats expmeta.dc3
.RE
Signed minimum value = 0x0      (0 dec)
.RE
Signed maximum value = 0x6e9    (1769 dec)
.RE
Unsigned minimum value = 0x0    (0 dec)
.RE
Unsigned maximum value = 0x6e9  (1769 dec)
.RE
Signed range = 0x6e9    (1769 dec)
.RE
Unsigned range = 0x6e9  (1769 dec)
.RE
.RE
Signed mean = 0x158     (344 dec)       (19.446%)
.RE
Unsigned mean = 0x158   (344 dec)       (19.446%)
.RE
Bits Stored = 16
.RE
Significant unsigned bits = 11
.RE
\ 
.RE
% dcstats impnone.dc3
.RE
Signed minimum value = 0xfa24   (64036 dec)
.RE
Signed maximum value = 0x638    (1592 dec)
.RE
Unsigned minimum value = 0x0    (0 dec)
.RE
Unsigned maximum value = 0xffff (65535 dec)
.RE
Signed range = 0xc14    (3092 dec)
.RE
Unsigned range = 0xffff (65535 dec)
.RE
Signed mean = 0xfec2    (65218 dec)     (38.2277%)
.RE
Unsigned mean = 0x1cdd  (7389 dec)      (11.2749%)
.RE
Bits Stored = 16
.RE
Significant positive bits = 11
.RE
Significant negative bits = 11
.RE
Offset to make unsigned in hex is = 0x5dc
.RE
Offset to make unsigned in dec is = 1500
.RE
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dchist(1) ,
.BR dcintro(1)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/dcfile/dcstats.cc
#include "attrmxls.h"
#include "attrothr.h"
#include "attrval.h"
#include "elmconst.h"
#include "mesgtext.h"
#include "dcopt.h"

static unsigned
howManySignificantBits(Uint16 test)
{
	unsigned bits;
	for (bits=0; test; test=test>>1,++bits);
	return bits;
}

static bool
dumpImageStatistics(ManagedAttributeList& list,TextOutputStream& log)
{
	bool success;

	Attribute *aBitsStored = list[TagFromName(BitsStored)];
	Uint16 vBitsStored=AttributeValue(aBitsStored);
	Attribute *aPixelRepresentation = list[TagFromName(PixelRepresentation)];
	Uint16 vPixelRepresentation=AttributeValue(aPixelRepresentation);
	Attribute *aLargestImagePixelValue = list[TagFromName(LargestImagePixelValue)];
	Uint16 vLargestImagePixelValue=AttributeValue(aLargestImagePixelValue);
	Attribute *aSmallestImagePixelValue = list[TagFromName(SmallestImagePixelValue)];
	Uint16 vSmallestImagePixelValue=AttributeValue(aSmallestImagePixelValue);

	Attribute *aRows = list[TagFromName(Rows)];
	if (!aRows)
		log << EMsgDC(MissingAttribute) << " - \"Rows\"" << endl;
	Uint16 vRows=AttributeValue(aRows);

	Attribute *aColumns = list[TagFromName(Columns)];
	if (!aColumns)
		log << EMsgDC(MissingAttribute) << " - \"Columns\"" << endl;
	Uint16 vColumns=AttributeValue(aColumns);

	Attribute *aNumberOfFrames = list[TagFromName(NumberOfFrames)];
	Uint16 vNumberOfFrames=AttributeValue(aNumberOfFrames,1);

	Attribute *aPixelData = list[TagFromName(PixelData)];
	if (!aPixelData)
		log << EMsgDC(MissingAttribute) << " - \"PixelData\"" << endl;

	if (!aPixelData || !aRows || !aColumns) {
		success=false;
	}
	else {
		SupplySourceFromAttribute sPixelData(aPixelData);
		class SourceBase<Uint16> *source = sPixelData.getSource();
		if (!source) {
			log << EMsgDC(BadAttributeValue)
			    << " - \"PixelData\""
			    << endl;
			success=false;
		}
		else {
			Int16  signedminval=Int16_MAX;
			Int16  signedmaxval=Int16_MIN;
			Uint16 unsignedminval=Uint16_MAX;
			Uint16 unsignedmaxval=0;
			Int32  signedtotal=0;
			Uint32 unsignedtotal=0;

			unsigned row=0;
			unsigned column=0;
			while (row < vRows*vNumberOfFrames && source->read()) {
				size_t n=source->getBufferCount();
				Assert(n);
				const Uint16 *buffer=source->getBuffer();
				Assert(buffer);
				while (n-- && row < vRows*vNumberOfFrames) {
					Uint16 value=*buffer++;
					if (Int16(value) < signedminval) signedminval=value;
					if (Int16(value) > signedmaxval) signedmaxval=value;
					if (value < unsignedminval) unsignedminval=value;
					if (value > unsignedmaxval) unsignedmaxval=value;
					unsignedtotal+=value;
					signedtotal+=Int16(value);
					if (++column >= vColumns) {
						column=0;
						++row;
					}
				}
			}
			//Assert(column == 0 && row == vRows*vNumberOfFrames);

			log << "Signed minimum value = " << hex << Uint16(signedminval) << dec << "\t(" << Uint16(signedminval) << " dec)" << endl;
			log << "Signed maximum value = " << hex << Uint16(signedmaxval) << dec << "\t(" << Uint16(signedmaxval) << " dec)" << endl;
			log << "Unsigned minimum value = " << hex << unsignedminval << dec << "\t(" << unsignedminval << " dec)" << endl;
			log << "Unsigned maximum value = " << hex << unsignedmaxval << dec << "\t(" << unsignedmaxval << " dec)" << endl;

			Uint16 signedrange=Uint16(signedmaxval-signedminval);
			Uint16 unsignedrange=unsignedmaxval-unsignedminval;

			log << "Signed range = "   << hex << signedrange   << dec << "\t(" << signedrange   << " dec)" << endl;
			log << "Unsigned range = " << hex << unsignedrange << dec << "\t(" << unsignedrange << " dec)" << endl;

			Int16 signedmean=Int16(signedtotal/(vColumns*vRows*vNumberOfFrames));
			Uint16 unsignedmean=Uint16(unsignedtotal/(vColumns*vRows*vNumberOfFrames));

			double signedmeanpercentofrange=signedrange ? ((signedmean-signedminval)*100.0/signedrange) : 0;
			double unsignedmeanpercentofrange=unsignedrange ? ((unsignedmean-unsignedminval)*100.0/unsignedrange) : 0;

			log << "Signed mean = "   << hex << Uint16(signedmean) << dec << "\t(" << Uint16(signedmean) << " dec)" << "\t(" << signedmeanpercentofrange   << "%)" << endl;
			log << "Unsigned mean = " << hex << unsignedmean       << dec << "\t(" << unsignedmean       << " dec)" << "\t(" << unsignedmeanpercentofrange << "%)" << endl;

			log << "Bits Stored = " << dec << vBitsStored << endl;
			if (signedrange < unsignedrange) {
				if (vPixelRepresentation == 0) {
					log << WMsgDC(PixelRepresentationIncorrect)
					    << " - " << MMsgDC(ShouldBe) << " " << MMsgDC(Signed)
					    << endl;
				}
				log << "Significant positive bits = " << dec << howManySignificantBits(signedmaxval) << endl;
				log << "Significant negative bits = " << dec << howManySignificantBits(-signedminval) << endl;
				log << "Offset to make unsigned in hex is = " << hex << Uint16(-signedminval) << endl;
				log << "Offset to make unsigned in dec is = " << dec << Uint16(-signedminval) << endl;
			}
			else if (signedrange > unsignedrange) {
				if (vPixelRepresentation == 1) {
					log << WMsgDC(PixelRepresentationIncorrect)
					    << " - " << MMsgDC(ShouldBe) << " " << MMsgDC(Unsigned)
					    << endl;
				}
				log << "Significant unsigned bits = " << dec << howManySignificantBits(unsignedmaxval) << endl;
			}
			else {
				if (vPixelRepresentation == 1) {
					log << WMsgDC(PixelRepresentationUnnecessary)
					    << " - " << MMsgDC(CouldBe) << " " << MMsgDC(Unsigned)
					    << endl;
				}
				log << "Significant unsigned bits = " << dec << howManySignificantBits(unsignedmaxval) << endl;
			}
			success=true;
		}
	}
	return success;
}

int
main(int argc, char *argv[])
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);

	bool verbose=options.get("verbose") || options.get("v");
	bool showfilename=options.get("filename");

	dicom_input_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);

	cerr << dicom_input_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();

	if (!dicom_input_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< " [-v|-verbose]"
			<< " [-filename]"
			<< " [" << MMsgDC(InputFile) << "]"
			<< " <" << MMsgDC(InputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);

	ManagedAttributeList list;

	bool success=true;
	TextOutputStream log(cerr);
	
	if (showfilename) {
		const char *filenameused = input_opener.getFilename();
		cerr << "Filename: \"" << (filenameused && strlen(filenameused) > 0 ? filenameused : "-") << "\"" << endl;
	}

	if (verbose) log << "******** While reading ... ********" << endl; 
	list.read(din,&log,verbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);

	const char *errors=list.errors();
	if (errors) log << errors << flush;
	if (!list.good()) {
		log << EMsgDC(DatasetReadFailed) << endl;
		success=false;
	}
	else {
		success=dumpImageStatistics(list,log);
	}

	return success ? 0 : 1;
}

